src/DataProvider/UserBookingItemDataProvider.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+// api/src/DataProvider/BlogPostItemDataProvider.php
+
+namespace App\DataProvider;
+
+use ApiPlatform\Core\DataProvider\ItemDataProviderInterface;
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use App\Entity\Main\UserBooking;
+use App\Service\MicrosoftGraphServiceInterface;
+use Exception;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+use Symfony\Component\Uid\Ulid;
+
+final class UserBookingItemDataProvider implements ItemDataProviderInterface, RestrictedDataProviderInterface
+{
+    public function __construct(private MicrosoftGraphServiceInterface $microsoftGraphService)
+    {
+    }
+
+    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
+    {
+        return UserBooking::class === $resourceClass;
+    }
+
+    public function getItem(string $resourceClass, $id, string $operationName = null, array $context = []): ?UserBooking
+    {
+        switch ($context['item_operation_name']) {
+            case 'get':
+                if (!isset($id)) {
+                    throw new BadRequestHttpException('Required booking id is not set');
+                }
+
+                $userId = ' ';
+                $userBookingResults = [$this->microsoftGraphService->getUserBooking($userId, $id)];
+                $userBooking = new UserBooking();
+
+                foreach ($userBookingResults as $userBookingResult) {
+                    $userBooking->id = Ulid::generate();
+                    $userBooking->hitId = $userBookingResult['id'] ?? '';
+                    $userBooking->summary = $userBookingResult['summary'] ?? '';
+                    $userBooking->subject = $userBookingResult['resource']['subject'] ?? '';
+                    $userBooking->start = new \DateTime($userBookingResult['start']['dateTime'], new \DateTimeZone($userBookingResult['start']['timeZone'])) ?? null;
+                    $userBooking->end = new \DateTime($userBookingResult['end']['dateTime'], new \DateTimeZone($userBookingResult['end']['timeZone'])) ?? null;
+                    $userBooking->iCalUId = $userBookingResult['iCalUId'];
+
+                    $bookingDetailsData = [$this->microsoftGraphService->getBookingDetails($userBookingResult['id'])];
+
+                    foreach ($bookingDetailsData as $bookingDetail) {
+                        $userBooking->displayName = $bookingDetail['location']['displayName'];
+                        $userBooking->body = $bookingDetail['body']['content'];
+                        continue;
+                    }
+                }
+
+                return $userBooking;
+            case 'delete':
```

Comment:
Shouldn't this be in a Data Persister instead of a Data Provider?

Reply:
According to the docs, a custom data provider can be used to get data from a seperate persistence layer, which in this case is our microsoft service. I think this is the way to do it, but not 100%. https://api-platform.com/docs/core/data-providers/

Reply:
But we are not getting... we are deleting

Reply:
ooo, this refers to the delete case. My bad